p2p/net/nat/internal/nat/natpmp.go

```diff
@@ -12,42 +12,48 @@ var (
 	_ NAT = (*natpmpNAT)(nil)
 )
 
-func discoverNATPMP(ctx context.Context) <-chan NAT {
-	res := make(chan NAT, 1)
-
+func discoverNATPMP(ctx context.Context) (NAT, error) {
 	ip, err := getDefaultGateway()
 	if err != nil {
-		return nil
+		return nil, err
 	}
 
-	go func() {
-		defer close(res)
-		// Unfortunately, we can't actually _stop_ the natpmp
-		// library. However, we can at least close _our_ channel
-		// and walk away.
-		select {
-		case client, ok := <-discoverNATPMPWithAddr(ip):
-			if ok {
-				res <- &natpmpNAT{client, ip, make(map[int]int)}
-			}
-		case <-ctx.Done():
-		}
-	}()
-	return res
-}
+	clientCh := make(chan *natpmp.Client)
+	errCh := make(chan error)
 
-func discoverNATPMPWithAddr(ip net.IP) <-chan *natpmp.Client {
-	res := make(chan *natpmp.Client, 1)
+	// We can't cancel the natpmp library, but we can at least still return
+	// on context cancellation by putting this in a goroutine
 	go func() {
-		defer close(res)
-		client := natpmp.NewClient(ip)
-		_, err := client.GetExternalAddress()
+		client, err := discoverNATPMPWithAddr(ctx, ip)
 		if err != nil {
+			errCh <- err
 			return
 		}
-		res <- client
+		clientCh <- client
```

Comment:
Should this be a buffered channel of size 1? Otherwise if the context gets cancelled this goroutine may hang forever.